File: Programmers/Python3/Level 2/방문 길이/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/49994

from collections import defaultdict


def solution(dirs):
    visited = defaultdict(set)
    x, y = 0, 0
    for dir in dirs:
        new_x, new_y = x, y
        if dir == "U":
            new_y += 1
        elif dir == "D":
            new_y -= 1
        elif dir == "R":
            new_x += 1
        else:
            new_x -= 1
        # 경로가 유효하다면 양 좌표가 시작점인 경로를 추가
        if -5 <= new_x <= 5 and -5 <= new_y <= 5:
            visited[(x, y)].add((new_x, new_y))
            visited[(new_x, new_y)].add((x, y))
            x, y = new_x, new_y
    # 양 좌표가 시작점인 경로를 추가했으므로 2로 나눔
    return sum(len(s) for s in visited.values()) // 2


def solution(dirs):  # 내 코드보다 짧은 코드, Ref: https://velog.io/@letgodchan0/%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%A8%B8%EC%8A%A4-%EB%B0%A9%EB%AC%B8-%EA%B8%B8%EC%9D%B4
    s = set()
    d = {'U': (0, 1), 'D': (0, -1), 'R': (1, 0), 'L': (-1, 0)}
    x, y = 0, 0
    for i in dirs:
        nx, ny = x + d[i][0], y + d[i][1]
        if -5 <= nx <= 5 and -5 <= ny <= 5:
            s.add((x, y, nx, ny))
            s.add((nx, ny, x, y))
            x, y = nx, ny
    return len(s)//2

File: Programmers/Javascript/Level 2/방문 길이/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/49994?language=javascript

/**
 * 해당 좌표가 보드안 좌표인지 알려줌
 * @param {number} x x좌표
 * @param {number} y y좌표
 * @return {boolean} 보드안의 좌표라면 true반환
 */
function validPoint(x, y) {
  return -5 <= x && x <= 5 && -5 <= y && y <= 5;
}

/**
 * 새로운 경로인지 확인하고 새로운 경로일시 기록
 * @param {number} x 출발 x좌표
 * @param {number} y 출발 y좌표
 * @param {number} newX 도착 x좌표
 * @param {number} newY 도착 y좌표
 * @return {boolean} 새로운 경로라면 true반환
 */
function isNewPath(x, y, newX, newY, prevPath) {
  const startPoint = `${x}, ${y}`;
  const endPoint = `${newX}, ${newY}`;

  if (!prevPath[startPoint]) {
    prevPath[startPoint] = new Set();
  }

  if (!prevPath[endPoint]) {
    prevPath[endPoint] = new Set();
  }

  if (prevPath[startPoint].has(endPoint)) {
    return false;
  }

  prevPath[startPoint].add(endPoint);
  prevPath[endPoint].add(startPoint);
  return true;
}

function solution(dirs) {
  const prevPath = {};
  let x = 0,
    y = 0,
    answer = 0;

  for (const command of dirs) {
    let newX = x,
      newY = y;
    switch (command) {
      case "U":
        newY += 1;
        break;
      case "L":
        newX -= 1;
        break;
      case "R":
        newX += 1;
        break;
      case "D":
        newY -= 1;
        break;
    }

    if (validPoint(newX, newY)) {
      if (isNewPath(x, y, newX, newY, prevPath)) {
        answer += 1;
      }
      x = newX;
      y = newY;
    }
  }
  return answer;
}
